Userland/Games/ColorLines/ColorLines.cpp

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ColorLines.h"
+#include "HueFilter.h"
+#include "Marble.h"
+#include "MarbleBoard.h"
+#include <LibConfig/Client.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Painter.h>
+
+ColorLines::BitmapArray ColorLines::build_marble_color_bitmaps()
+{
+    auto marble_bitmap = MUST(Gfx::Bitmap::try_load_from_file("/res/icons/colorlines/colorlines.png"sv));
+    float const hue_degrees[Marble::number_of_colors] = {
+        0,   // Red
+        45,  // Brown/Yellow
+        90,  // Green
+        180, // Cyan
+        225, // Blue
+        300  // Purple
+    };
+    BitmapArray colored_bitmaps;
+    colored_bitmaps.ensure_capacity(Marble::number_of_colors);
+    for (int i = 0; i < Marble::number_of_colors; ++i) {
+        auto bitmap = MUST(marble_bitmap->clone());
+        HueFilter filter { hue_degrees[i] };
+        filter.apply(*bitmap, bitmap->rect(), *marble_bitmap, marble_bitmap->rect());
+        colored_bitmaps.append(bitmap);
+    }
+    return colored_bitmaps;
+}
+
+ColorLines::BitmapArray ColorLines::build_marble_trace_bitmaps()
+{
+    auto trace_bitmap = MUST(Gfx::Bitmap::try_load_from_file("/res/emoji/U+1F43E.png"sv));
+    BitmapArray result;
+    result.ensure_capacity(4);
+    result.append(trace_bitmap);
+    result.append(MUST(result.last()->rotated(Gfx::RotationDirection::Clockwise)));
+    result.append(MUST(result.last()->rotated(Gfx::RotationDirection::Clockwise)));
+    result.append(MUST(result.last()->rotated(Gfx::RotationDirection::Clockwise)));
+    return result;
+}
+
+ColorLines::ColorLines(StringView app_name)
+    : m_app_name { app_name }
+    , m_game_state { GameState::Idle }
+    , m_board { make<MarbleBoard>() }
+    , m_marble_bitmaps { build_marble_color_bitmaps() }
+    , m_trace_bitmaps { build_marble_trace_bitmaps() }
+    , m_score_font { Gfx::BitmapFont::load_from_file("/res/fonts/MarietaBold24.font") }
+{
+    VERIFY(m_marble_bitmaps.size() == Marble::number_of_colors);
+    set_font(Gfx::FontDatabase::default_fixed_width_font().bold_variant());
+    m_high_score = Config::read_i32(m_app_name, m_app_name, "HighScore"sv, 0);
+    reset();
+}
+
+void ColorLines::reset()
+{
+    set_game_state(GameState::StartingGame);
+}
+
+void ColorLines::mousedown_event(GUI::MouseEvent& event)
+{
+    if (m_game_state != GameState::Idle && m_game_state != GameState::MarbleSelected)
+        return;
+    auto const event_position = event.position().translated(-frame_inner_rect().x(), -frame_inner_rect().y() - board_vertical_margin);
+    if (event_position.x() < 0 || event_position.y() < 0)
+        return;
+    auto const clicked_cell = Point { event_position.x() / board_cell_dimension.width(), event_position.y() / board_cell_dimension.height() };
```

Comment:
There's quite a few functions that could use some newlines -- here's just one example, where I'd say putting a newline after an `if` statement would be desirable.